src/metabase/api/dataset.clj

```diff
@@ -77,15 +78,45 @@
   (or (get-in ex/export-formats [export-format :context])
       (throw (Exception. (str "Invalid export format: " export-format)))))
 
+(def ^:private date-only-formatter (tformat/formatters :date))
+(def ^:private iso-date-formatter (tformat/formatters :date-time))
+
+(defn- datetime-str->date
+  "Parse `date-str` as an iso formatted timestamp and emit the date-only portion of that timestamp"
+  [date-str]
+  (->> date-str
+       (tformat/parse iso-date-formatter)
+       (tformat/unparse date-only-formatter)))
+
+(defn- swap-date-columns [date-col-indexes]
+  (fn [row]
+    (reduce (fn [acc idx]
+              (update acc idx datetime-str->date)) row date-col-indexes)))
+
+(defn- date-column-indexes
+  "Given `column-metadata` find the `:type/Date` columns"
+  [column-metadata]
+  (transduce (comp (map-indexed (fn [idx col-map] [idx (:base_type col-map)]))
+                   (filter (fn [[idx base-type]] (isa? base-type :type/Date)))
+                   (map first))
+             conj [] column-metadata))
+
+(defn- maybe-modify-date-values [column-metadata rows]
+  (let [date-indexes (date-column-indexes column-metadata)]
+    (if (seq date-indexes)
+      ;; Not sure why, but rows aren't vectors, they're lists which makes updating difficult
+      (map (comp (swap-date-columns date-indexes) vec) rows)
```

Comment:
This is also not very efficient. In some areas of the codebase, rows are vectors, not generic seqs. I'm going to see if I can find out where this is getting converted from a vector and fix it there so there's not a need to convert it here.

Reply:
Created this to address the problem separately https://github.com/metabase/metabase/issues/6723

Reply:
Why do you need to convert it to a vector here?

Reply:
I guess I don't need to, but our results format is setup for index based access. We have column metadata in the `:cols` key and the only way to connect a cell in the resultset to a column in that `:cols` key is by index. So in this PR I find out the index of the `:type/Date` columns in `:cols` and then need to modify those date columns in every row of the resultset. Vectors are very efficient and easy to update when you know the index. Generic seqs don't support index based access and the code would be more complicated to update the given number of cells. We would go from
``` clojure 
(reduce (fn [acc idx]
              (update acc idx datetime-str->date)) row date-col-indexes)
```
to
``` clojure
(reduce (fn [acc index]
          (concat (take (dec index) acc)
                  (cons (datetime-str->date (nth acc index))
                        (drop (inc index)))))
        row date-col-indexes)
```
Where we're traversing cells in the row 3 times for every date updated. We could probably get more clever, maybe `split-at` or something, which will create more garbage to collect. But it's working around what I see is the issue of individual rows not being a vector. There are some places in the codebase where we need to modify rows as they flow out, remapping is another example. Having them be a vector makes it easy and efficient to make those changes.

One alternative could be to write the seq knitting code above and once we have vectors flowing through, change it back.